wsgi.py

```diff
@@ -6,255 +6,348 @@
+@application.route("/projects/<project_uuid>", methods=["PUT"])
+@application.route("/projects/<project_uuid>/owner/<user_name>", methods=["PUT"])
+@auth.login_required
 def create_moc_project(project_uuid, user_name=None):
-    (token, openshift_url) = get_token_and_url()
+    shift = get_openshift()
     # first check the project_name is a valid openshift project name
-    suggested_project_name = cnvt_project_name(project_uuid)
-    if(project_uuid != suggested_project_name):
+    suggested_project_name = shift.cnvt_project_name(project_uuid)
+    if project_uuid != suggested_project_name:
         # future work, handel colisons by suggesting a different valid
         # project name
         return Response(
-            response=json.dumps({"msg":"ERROR: project name must match regex '[a-z0-9]([-a-z0-9]*[a-z0-9])?'", "suggested name": suggested_project_name }),
+            response=json.dumps(
+                {
+                    "msg": "ERROR: project name must match regex '[a-z0-9]([-a-z0-9]*[a-z0-9])?'",
+                    "suggested name": suggested_project_name,
+                }
+            ),
             status=400,
-            mimetype='application/json'
-            )
-    if(not exists_openshift_project(token, openshift_url, project_uuid)):
-        project_name=project_uuid
-        if("Content-Length" in request.headers):
-            req_json=request.get_json(force=True)
-            if("displayName" in req_json):
-                project_name=req_json["displayName"]
-            application.logger.debug("create project json: "+project_name)
+            mimetype="application/json",
+        )
+    if not shift.project_exists(project_uuid):
+        project_name = project_uuid
+        if "Content-Length" in request.headers:
+            req_json = request.get_json(force=True)
+            if "displayName" in req_json:
+                project_name = req_json["displayName"]
+            application.logger.debug("create project json: " + project_name)
         else:
             application.logger.debug("create project json: None")
 
-        r = create_openshift_project(token, openshift_url, project_uuid, project_name, user_name)
-        if(r.status_code == 200 or r.status_code == 201):
+        r = shift.create_project(project_uuid, project_name, user_name)
+        if r.status_code == 200 or r.status_code == 201:
             return Response(
-                response=json.dumps({"msg": "project created (" + project_uuid +")" }),
+                response=json.dumps({"msg": "project created (" + project_uuid + ")"}),
                 status=200,
-                mimetype='application/json'
-                )
+                mimetype="application/json",
+            )
         return Response(
-            response=json.dumps({"msg": "project unabled to be created (" + project_uuid +")" }),
+            response=json.dumps(
+                {"msg": "project unabled to be created (" + project_uuid + ")"}
```

Comment:
When failing because of an http error, it's helpful to return details that can be used to debug the problem (or at least log them on the console). E.g.:

```
application.logger.debug("create project failed: {}".format(r.text)
```

Reply:
Yes, however, I also don't want to leak much information to the end user.

So the question becomes what should be logged.  Should all entrances and exits from functions be logged?

Reply:
I wouldn't log all exits and entrances, but if we get a failure from the openshift api I would log *that*. I don't mean expose it to the end user, I mean make it available in the application logs.